src/bank_checkpoint.rs

```diff
@@ -0,0 +1,348 @@
+    }
+
+    pub fn first_err(results: &[Result<()>]) -> Result<()> {
+        for r in results {
+            r.clone()?;
+        }
+        Ok(())
+    }
+
+    pub fn commit_transactions(
+        &self,
+        subscritpions: &Option<Arc<RpcSubscriptions>>,
+        txs: &[Transaction],
+        loaded_accounts: &[Result<(InstructionAccounts, InstructionLoaders)>],
+        executed: &[Result<()>],
+    ) {
+        assert!(!self.frozen());
+        let now = Instant::now();
+        self.accounts
+            .store_accounts(true, txs, executed, loaded_accounts);
+
+        // Check account subscriptions and send notifications
+        if let Some(subs) = subscritpions {
+            Self::send_account_notifications(subs, txs, executed, loaded_accounts);
+        }
+
+        // once committed there is no way to unroll
+        let write_elapsed = now.elapsed();
+        debug!(
+            "store: {}us txs_len={}",
+            duration_as_us(&write_elapsed),
+            txs.len(),
+        );
+        self.update_transaction_statuses(txs, &executed);
+        if let Some(subs) = subscritpions {
+            Self::update_subscriptions(subs, txs, &executed);
+        }
+    }
+    fn send_account_notifications(
+        subscriptions: &RpcSubscriptions,
+        txs: &[Transaction],
+        res: &[Result<()>],
+        loaded: &[Result<(InstructionAccounts, InstructionLoaders)>],
+    ) {
+        for (i, raccs) in loaded.iter().enumerate() {
+            if res[i].is_err() || raccs.is_err() {
+                continue;
+            }
+
+            let tx = &txs[i];
+            let accs = raccs.as_ref().unwrap();
+            for (key, account) in tx.account_keys.iter().zip(accs.0.iter()) {
+                subscriptions.check_account(&key, account);
+            }
+        }
+    }
+    fn update_subscriptions(
+        subscriptions: &RpcSubscriptions,
+        txs: &[Transaction],
+        res: &[Result<()>],
+    ) {
+        for (i, tx) in txs.iter().enumerate() {
+            subscriptions.check_signature(&tx.signatures[0], &res[i]);
+        }
+    }
+
+    fn update_transaction_statuses(&self, txs: &[Transaction], res: &[Result<()>]) {
+        assert!(!self.frozen());
+        let mut status_cache = self.status_cache.write().unwrap();
+        for (i, tx) in txs.iter().enumerate() {
+            match &res[i] {
+                Ok(_) => status_cache.add(&tx.signatures[0]),
+                Err(BankError::LastIdNotFound) => (),
+                Err(BankError::DuplicateSignature) => (),
+                Err(BankError::AccountNotFound) => (),
+                Err(e) => {
+                    status_cache.add(&tx.signatures[0]);
+                    status_cache.save_failure_status(&tx.signatures[0], e.clone());
+                }
+            }
+        }
+    }
+
+    pub fn hash_internal_state(&self) -> Hash {
+        self.accounts.hash_internal_state()
+    }
+    pub fn set_genesis_last_id(&self, last_id: &Hash) {
```

Comment:
That's a strange one to see a method for. How would it ever change? How come that's not passed into the constructor?

Reply:
@mvines

Reply:
Are you asking me something?

Reply:
that's my code, and it populates a valid last_id without requiring a "tick" in genesis, which screws up schedules  (off by one)